### PR TITLE
Support SkipLocalsInitAttribute in local functions

### PR DESCRIPTION
**Pull request by RikkiGibson**
Related to #38801

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberMethodSymbol.cs [hunk 1]
```diff
@@ -367,15 +367,6 @@ public override Symbol AssociatedSymbol
             }
         }
 
-        public override bool AreLocalsZeroed
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberMethodSymbol.cs]: Deleted redundant implementation that was inadvertently left around after a merge

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs [hunk 2]
```diff
@@ -345,13 +345,12 @@ internal sealed override CSharpAttributeData EarlyDecodeWellKnownAttribute(ref E
             return base.EarlyDecodeWellKnownAttribute(ref arguments);
         }
 
-        // PROTOTYPE(local-function-attributes): test local function with SkipLocalsInit
         public override bool AreLocalsZeroed
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs]: >override [](start = 15, length = 8)

Can we seal this implementation?

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_WellKnownAttributes.cs [hunk 3]
```diff
@@ -9413,6 +9413,158 @@ void LF()
...
+                x = x + x + x;
+            };
+        }
+
+        System.Action L = () =>
+        {
+            int y = 4;
+            y = y + y + y;
+
+            void LF()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_WellKnownAttributes.cs]: >void LF() [](start = 12, length = 9)

It looks like we are missing a positive test for a scenario when a local function is nested into a lambda.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs [hunk 4]
```diff
@@ -33,6 +33,27 @@ protected static void ReportBadRefToken(TypeSyntax returnTypeSyntax, DiagnosticB
             }
         }
 
+        protected bool AreContainingSymbolLocalsZeroed
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs]: You could do 
```C#
protected bool AreContainingSymbolLocalsZeroed => ContainingSymbol switch
{
    SourceMethodSymbol m => m.AreLocalsZeroed,
    SourceNamedTypeSymbol t => t.AreLocalsZeroed,
    _ => true
};
```

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_WellKnownAttributes.cs [hunk 5]
```diff
@@ -9413,6 +9413,158 @@ void LF()
...
+    {
+    }
+}
+
+public class C
+{
+    public void M()
+    {
+        [System.Runtime.CompilerServices.SkipLocalsInitAttribute]
+        void F()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_WellKnownAttributes.cs]: >void F() [](start = 8, length = 8)

I think we should also test scenarios when the attribute is applied to a containing member (including events/properties if allowed), or containing type, or containing module.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs]: I think we need to override it in SynthesizedMethodBaseSymbol.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs]: I think we need to override it in SynthesizedMethodBaseSymbol.

---
In reply to: [371410335](https://github.com/dotnet/roslyn/pull/41183#discussion_r371410335) [](ancestors = 371410335)

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs]: > I think we need to override it in SynthesizedMethodBaseSymbol.

SynthesizedMethodBaseSymbol inherits from SourceMemberMethodSymbol

---
In reply to: [371421097](https://github.com/dotnet/roslyn/pull/41183#discussion_r371421097) [](ancestors = 371421097,371410335)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs (hunk 4 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs]: Does this handle the scenario where the ContainingSymbol is a SourceFieldSymbolFromDeclarator and the containing type has SkipLocalsInitAttribute?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs]: I think when I was triaging this I was surprised to find this as the ContainingSymbol for some scenario or other, but I tried to create a test that would create a bad behavior here and couldn't do it, so I'm good with taking your suggestion.

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_WellKnownAttributes.cs (hunk 5 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_WellKnownAttributes.cs]: Some of these scenarios already had tests, such as `SkipLocalsInitAttributeOnMethodPropagatesToLocalFunction`, but I added tests for the other scenarios you mentioned.

---
In reply to: [371418225](https://github.com/dotnet/roslyn/pull/41183#discussion_r371418225) [](ancestors = 371418225)

**Comment by RikkiGibson** [comment]
@dotnet/roslyn-compiler Please review

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_WellKnownAttributes.cs [hunk 6]
```diff
@@ -10789,6 +11284,45 @@ static int M(out int x)
             Assert.False(verifier.HasLocalsInit("C..ctor"));
         }
 
+        [Fact]
+        public void SkipLocalsInitFieldInitializer()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_WellKnownAttributes.cs]: >SkipLocalsInitFieldInitializer [](start = 20, length = 30)

It doesn't look like this test actually tests anything in field initializer. However, I think this is an interesting scenario to test. We could have a lambda with a local function within a field initializer. We also could have two constructors: one with an attribute and one without. It feels like the attributes on the constructors shouldn't have any effect on the lambda and the local function. But the attribute on the type/module should have an effect. Please add a test like this.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_WellKnownAttributes.cs]: What's a more appropriate term for this test? "ImplicitConstructor"?

---
In reply to: [371973916](https://github.com/dotnet/roslyn/pull/41183#discussion_r371973916) [](ancestors = 371973916)

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_WellKnownAttributes.cs]: > What's a more appropriate term for this test? "ImplicitConstructor"?

Yes, I think this term will describe the purpose of the test more accurately.

---
In reply to: [371989150](https://github.com/dotnet/roslyn/pull/41183#discussion_r371989150) [](ancestors = 371989150,371973916)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs (hunk 4 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs]: Aleksey came up with a test case involving a field initializer containing a lambda, where the ContainingSymbol is a FieldSymbol and we have to instead look up to the ContainingType. So I reverted the suggestion.

Comment (agocke) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs]: Were you thinking of a lambda inside a field initializer?

Comment (agocke) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs]: I could see that as being a reasonable scenario that would be broken with my code.

**Pull request merged**